src/Tap.t.sol

```diff
@@ -0,0 +1,273 @@
+
+    function setUp() public {
+        hevm = Hevm(address(CHEAT_CODE));
+
+        me = address(this);
+
+        vat = new TestVat();
+        flapper = new TestFlapper(address(vat));
+        vow = new TestVow(address(vat), address(flapper));
+
+        dai = DaiAbstract(address(new Dai(0)));
+        daiJoin = DaiJoinAbstract(address(new DaiJoin(address(vat), address(dai))));
+        vat.rely(address(daiJoin));
+        dai.rely(address(daiJoin));
+
+        keg = KegAbstract(address(new Keg(address(dai))));
+
+        tap = new Tap(keg, daiJoin, address(vow), "operations", WAD / (1 days));
+        vat.rely(address(tap));
+
+        wallet1 = new Wallet();
+        wallet2 = new Wallet();
+        wallet3 = new Wallet();
+
+        address[] memory wallets = new address[](2);
+        wallets[0] = address(wallet1);
+        wallets[1] = address(wallet2);
+        uint256[] memory amts = new uint256[](2);
+        amts[0] = 0.25 ether;   // 25% split
+        amts[1] = 0.75 ether;   // 75% split
+        bytes32 flight = "operations";
+
+        keg.seat(flight, wallets, amts);
+
+    }
+
+
+    // file("flight")
+    function test_flight() public {
+        assertEq(tap.flight(), "operations");
+
+        tap.file("flight", "opt");
+
+        assertEq(tap.flight(), "opt");
+    }
+
+    function testFail_flight() public {
+        tap.deny(me);
+        tap.file("flight", "opt");
+    }
+
+    function testFail_flight_with_time_effect() public {
+        hevm.warp(1 days);
+        tap.file("flight", "opt");
+    }
+
+    function test_flight_with_time_effect_after_pump() public {
+        hevm.warp(1 days);
+        tap.pump();
+        tap.file("flight", "opt");
+
+        assertEq(tap.flight(), "opt");
+    }
+
+    // file("rate")
+    // TODO  need rate definition
+    function test_rate() public {
+        // assertEq(tap.rate, 0); // TODO need rate definition
+        tap.file("rate", 100000);
+
+        assertEq(tap.rate(), 100000);
+    }
+
+    function testFail_rate() public {
+        tap.deny(me);
+        tap.file("rate", uint256(0));
+    }
+
+    function testFail_rate_with_time_effect() public {
+        hevm.warp(1 days);
+        tap.file("rate", uint256(0));
+    }
+
+    function test_rate_with_time_effect_after_pump() public {
+        // assertEq(tap.rate, 0); // TODO need rate definition
+        hevm.warp(1 days);
+        tap.pump();
+        tap.file("rate", uint256(0));
+
+        assertEq(tap.rate(), uint256(0));
+    }
+
+
+    // pump()
+    function test_pump() public {
+        hevm.warp(1 seconds);
+        tap.pump();
+    }
+
+    function testFail_pump_call_twice_same_block() public {
+        hevm.warp(1 seconds);
+        tap.pump();
+        tap.pump();
+    }
+
+    function testFail_pump_without_no_monies_to_send() public {
+        tap.file("rate", uint256(0));
+        hevm.warp(1 seconds);
+        tap.pump();
+    }
+
+}
```

Comment:
Same as the other test. This should just be merged into the Keg.t.sol.

Reply:
To me this makes sense for a larger project, but the keg is really small and so I don't think we need this level of separation.

Reply:
Ha ha ha,
Ok here is my point those tests are sort of value less for the general high level project understanding however if the `auth` is removed/not working they are just the most needed test.

Here we go I can put this test back to the keg.t.sol or remove them.

Reply:
I suppose it's not a big deal to just leave that as is.